### ISSUE TITLE
Bug: [no-floating-promises] should not be tricked by spread args 

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#9808]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=MYewdgzgLgBApgWwA5QJ4CU4RAGwG5wBOEAXDAIKGECGqAPABQCUMAvAHwx4gCWAJp1YwA2gF0A3ACgACoRAIeEOADpCcAFZxgUZsqgALOGAbLTiFBiy4CxADQxTy82kzZ8RCE3FA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooAd37QmkcGAC%2BIDUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
const emptyResolvers: Array<() => void> = [];
Promise.reject().then(...emptyResolvers, ...emptyResolvers);
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-floating-promises": "warn"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

I expected to see an error that a floating promise is created, since no handler is passed.

### Actual Result

No error.

### Additional Info

This is a spinoff of https://github.com/typescript-eslint/typescript-eslint/issues/8958#issuecomment-2067713381, and relates to https://github.com/typescript-eslint/typescript-eslint/pull/9059#discussion_r1707667740. 

Implementation-wise, I'm proposing we take the dead-simple approach and assume that if spread args are passed, the promise isn't handled. I have code for this change ready to go.

**Comment by Josh-Cena** [comment]
-1. This is a bad assumption and if someone really does this (I occasionally do, especially for things like `const handlers = [() => true, () => false]; FS.access(path).then(...handlers)`), we should be optimistic and assume they know what they are doing.

**Comment by kirkwaiblinger** [comment]
@Josh-Cena Yeah, I'm fine with going the optimistic route as well, as long as we're consistent. But I don't like that 
```ts
const emptyResolvers: Array<() => void> = [];
// this flags
Promise.reject().then(...emptyResolvers);
// but this doesn't
Promise.reject().then(...emptyResolvers, ...emptyResolvers);
```
The rule is just oblivious right now to the fact that spread args are present at all.

**Comment by kirkwaiblinger** [comment]
PS - Note that the very case that you bring up currently _does_ flag, so we'd have to make a change in order to change that behavior.

**Comment by Josh-Cena** [comment]
Ah I see what you mean. My personal vote is that spread elements should either be consistently treated as zero-length or infinite-length, but not implicitly length-1. I would prefer infinite-length.

**Comment by kirkwaiblinger** [comment]
> My personal vote is that spread elements should either be consistently treated as zero-length or infinite-length, but not implicitly length-1. I would prefer infinite-length.

Hmm, well, this terminology brings up the question - should the following (perfectly valid according to the types) code flag? 

```ts
declare const arrayOfUndefined: Array<undefined>;
Promise.reject().then(...arrayOfUndefined);
```

| spread args strategy | should it flag? | why? |
| --- | --- | --- |
| "pessimistic" (my original proposal) | YES | we see spread args and assume the worst (from syntax alone, no type checking) |
| "optimistic" | NO | we see spread args and assume the user knows what they're doing (from syntax alone, no type checking) | 
| "zero length" | YES | spread args are ignored, and there's nothing afterwards | 
| "infinite length" | YES | spread arg is assumed to be infinitely wide, _but the type info indicates it is not a function_ |

And then what about this case? (which is in fact handled)
```ts
declare const tupleOfUndefined: [undefined];
Proimse.reject().then(...tupleOfUndefined, () => { /* valid handler */ });
```

| spread args strategy | should it flag? | why? |
| --- | --- | --- |
| "pessimistic"  | YES | we see spread args and assume the worst (from syntax alone, no type checking) |
| "optimistic" | NO | we see spread args and assume the user knows what they're doing (from syntax alone, no type checking) | 
| "zero length" | YES | spread args are ignored, even though it's a fixed length tuple type, making the valid catch handler get analyzed as though it were the fulfillment handler | 
| "infinite length" | YES | spread arg is assumed to be infinitely wide _despite being a fixed-length tuple_, and the type info indicates it is not a function. |

```ts
declare const tupleOfUndefined: [undefined];
Proimse.reject().then(...tupleOfUndefined, undefined, () => { /* valid handler */ });
```

| spread args strategy | should it flag? | why? |
| --- | --- | --- |
| "pessimistic"  | YES | we see spread args and assume the worst (from syntax alone, no type checking) |
| "optimistic" | NO | we see spread args and assume the user knows what they're doing (from syntax alone, no type checking) | 
| "zero length" | NO | spread args are ignored, even though it's a fixed length tuple type, shifting the valid handler into the catch position 😥  | 
| "infinite length" | YES | spread arg is assumed to be infinitely wide _despite being a fixed-length tuple_, and the type info indicates it is not a function. |

.... and we can think of several other perverse edge cases along the lines of tuples, and args after spread.

***

This is to say, within the terminology I just made up here, I think we'd be better off choosing between a syntax-analysis-only "pessimistic" or "optimistic" strategy, rather than an infinite- or zero-length-with-type-checking approach.

**Comment by bradzacher** [comment]
[This sourcegraph query](https://sourcegraph.com/search?q=context:global+/%5C.then%5C%28%5C.%5C.%5C.%5Cw/+lang:TypeScript&patternType=keyword&sm=0) reports 156 results for `/\.then\(\.\.\.\w/` and 115 of those are in one repo's tests.

So:
- using `.then` on a top-level floating promise is already the rarer case
- spreading into `.then` this is a 1-in-a-billion usecase on top of that

I don't think it's worth putting any effort into this case, personally.
My 2c: I think at most we just document the limitation and move on.

**Comment by kirkwaiblinger** [comment]
Yeah, I mean, similar sentiment from me, except I was thinking it would be easier to just short circuit the rule logic (using the existing PR) rather than write coherent documentation 😂
